### ISSUE TITLE
remove_background not producing sensible output with 2D EELS map.

### ISSUE THREAD

**Issue by ZGainsforth** [hyperspy/hyperspy#2129]
I'm attempting to do a fourier_ratio_deconvolution between two EELS cubes.  One includes the ZLP and LL region and the other includes Fe-L core loss.  I run the following code:

```
import numpy as np
import matplotlib.pyplot as plt
import hyperspy.api as hs
import os

ZLP = hs.load('007-EELS-ZLP-45pA.emi')
ZLP.set_signal_type('EELS')
ZLP.set_microscope_parameters(beam_energy=300, convergence_angle=2, collection_angle=30)
FeL = hs.load('008-EELS_Map_core-360pA.emi')
FeL.set_signal_type('EELS')
FeL.set_microscope_parameters(beam_energy=300, convergence_angle=2, collection_angle=30)
print(f'ZLP shape is: {ZLP.data.shape}')
print(f'FeL shape is: {FeL.data.shape}')

ZLP.align_zero_loss_peak()

FeSub = FeL.remove_background(signal_range=(670., 695.), background_type='PowerLaw')
# FeSub.set_signal_type('EELS')
# FeDeconv = FeSub.fourier_ratio_deconvolution(ll=ZLP)

ZLP.plot()
FeL.plot()
FeSub.plot()
plt.show()
```

It produces the following output:

```
ZLP shape is: (35, 35, 2048)
FeL shape is: (35, 35, 2048)

Initial ZLP position statistics
-------------------------------
Summary statistics
------------------
mean:	0.0235
std:	0.146

min:	-0.8
Q1:	0
median:	0
Q3:	0.2
max:	0.4
HBox(children=(IntProgress(value=0, max=1225), HTML(value='')))

HBox(children=(IntProgress(value=0, max=1225), HTML(value='')))

HBox(children=(IntProgress(value=0, max=1225), HTML(value='')))

```

<img width="1923" alt="reproducebugoutput" src="https://user-images.githubusercontent.com/6864619/52254442-a3e57080-28c2-11e9-8aec-d3a70ae4a90e.png">

As you can see the low loss and core loss are OK, but the background subtraction is not.  I have tried a few different ranges on signal_range and it produces apparently the same output.  Also, I tried a polynomial background and it produced:

```
Traceback (most recent call last):
  File "ReproduceBug.py", line 17, in <module>
    FeSub = FeL.remove_background(signal_range=(670., 695.), background_type='Polynomial', polynomial_order=1)
  File "/Users/Zack/anaconda3/envs/conda37/lib/python3.7/site-packages/hyperspy/_signals/signal1d.py", line 1099, in remove_background
    show_progressbar=show_progressbar)
  File "/Users/Zack/anaconda3/envs/conda37/lib/python3.7/site-packages/hyperspy/_signals/signal1d.py", line 1040, in _remove_background_cli
    background_estimator.function_nd(axis))
ValueError: operands could not be broadcast together with shapes (35,35,2048) (35,2,2048)
```

I'm using hyperspy 1.4.1 on the conda 4.6.2 (just downloaded from continuum.io tonight).

```
# packages in environment at /Users/Zack/anaconda3/envs/conda37:
#
# Name                    Version                   Build  Channel
hyperspy                  1.4.1                 py37_1002    conda-forge
hyperspy-base             1.4.1           py37h1de35cc_1002    conda-forge
hyperspy-gui-ipywidgets   1.1.1                      py_1    conda-forge
hyperspy-gui-traitsui     1.1.1                 py37_1000    conda-forge
```

Is there a way to get this to work?

Thanks,

Zack

If you would like I can send the EELS cubes.  They are about 10 MB in a zip.

**Comment by ericpre** [comment]
They may be two different issue here, can you try with `fast=False` for the `remove_background` method? It may fix both issues.
What happen if you remove the background interactively without specifying `signal_range=(670., 695.)`? You will need 

The second issue with the polynomial background is the same as the one mentioned in #2114.

**Comment by ZGainsforth** [comment]
Yay!  "fast=False" works:

<img width="640" alt="image" src="https://user-images.githubusercontent.com/6864619/52272712-73222d00-28fc-11e9-9b5b-96f415e13499.png">

And  "fast=False" also makes a polynomial background subtraction happy.

When I take out the signal_range in an attempt to activate interactive mode it seems to need 'Power Law' instead of 'PowerLaw' and even then errors out.  So perhaps I'm not doing that part right:

`FeSub = FeL.remove_background(background_type='PowerLaw', fast=False)`

results in:

```
Traceback (most recent call last):
  File "ReproduceBug.py", line 18, in <module>
    FeSub = FeL.remove_background(background_type='PowerLaw', fast=False)
  File "/Users/Zack/anaconda3/envs/hyperspy37/lib/python3.7/site-packages/hyperspy/_signals/signal1d.py", line 1077, in remove_background
    zero_fill=zero_fill)
  File "/Users/Zack/anaconda3/envs/hyperspy37/lib/python3.7/site-packages/hyperspy/signal_tools.py", line 678, in __init__
    self.background_type = background_type
  File "/Users/Zack/anaconda3/envs/hyperspy37/lib/python3.7/site-packages/traits/trait_handlers.py", line 236, in error
    object, name, self.full_info(object, name, value), value
traits.trait_errors.TraitError: The 'background_type' trait of a BackgroundRemoval instance must be 'Power Law' or 'Gaussian' or 'Offset' or 'Polynomial', but a value of 'PowerLaw' <class 'str'> was specified.
```

and

`FeSub = FeL.remove_background(background_type='Power Law', fast=False)`

results in:

```
VBox(children=(FloatText(value=nan, description='Left', disabled=True), FloatText(value=nan, description='Right', disabled=True), Dropdown(description='Background type', options=('Power Law', 'Gaussian', 'Offset', 'Polynomial'), value='Power Law'), IntText(value=2, description='Polynomial order', layout=Layout(display='none')), Checkbox(value=False, description='Fast'), Checkbox(value=False, description='Zero Fill'), Accordion(children=(HTML(value='Click on the signal figure and drag to the right to select a range. Press `Apply` to remove the background in the whole dataset. If "Fast" is checked, the background parameters are estimated using a fast (analytical) method that can compromise accuracy. When unchecked, non-linear least squares is employed instead. If "Zero Fill" is checked, all the channels prior to the fitting region will be set to zero. Otherwise the background subtraction will be performed in the pre-fitting region as well.'),), _titles={'0': 'Help'}), HBox(children=(Button(description='Apply', style=ButtonStyle(), tooltip='Remove the background in the whole dataset.'), Button(description='Close', style=ButtonStyle(), tooltip='Close widget and remove span selector from the signal figure.')))))
Traceback (most recent call last):
  File "ReproduceBug.py", line 25, in <module>
    FeSub.plot()
AttributeError: 'NoneType' object has no attribute 'plot'
```

But anyway,  "fast=False" seems to be a suitable workaround.  Thanks!

**Comment by ericpre** [comment]
This looks to me like an installation issue.

Can you try to make a new environment:
```
conda create -n test python
source activate test
conda install hyperspy -c conda-forge
```

And test it again?

**Comment by ZGainsforth** [comment]
I tried it again in the test environment and I get the same output.  This is run from a terminal (iTerm2, OSX), Qt5Agg backend.  Not sure if that matters.

1) OK if I specify range and fast-False
2) Asks for "Power Law" if I remove range.
3) Errors if I say "Power Law"

`> conda info`

```
     active environment : test
    active env location : /Users/Zack/anaconda3/envs/test
            shell level : 4
       user config file : /Users/Zack/.condarc
 populated config files :
          conda version : 4.6.2
    conda-build version : 3.17.6
         python version : 3.7.1.final.0
       base environment : /Users/Zack/anaconda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/osx-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/Zack/anaconda3/pkgs
                          /Users/Zack/.conda/pkgs
       envs directories : /Users/Zack/anaconda3/envs
                          /Users/Zack/.conda/envs
               platform : osx-64
             user-agent : conda/4.6.2 requests/2.21.0 CPython/3.7.1 Darwin/18.2.0 OSX/10.14.2
                UID:GID : 501:20
             netrc file : None
           offline mode : False
```

`> conda list`

```
# packages in environment at /Users/Zack/anaconda3/envs/test:
#
# Name                    Version                   Build  Channel
appnope                   0.1.0                 py37_1000    conda-forge
asn1crypto                0.24.0                py37_1003    conda-forge
attrs                     18.2.0                     py_0    conda-forge
backcall                  0.1.0                      py_0    conda-forge
blas                      1.0                         mkl
bokeh                     1.0.4                 py37_1000    conda-forge
ca-certificates           2018.11.29           ha4d7672_0    conda-forge
certifi                   2018.11.29            py37_1000    conda-forge
cffi                      1.11.5          py37h342bebf_1001    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
click                     7.0                        py_0    conda-forge
cloudpickle               0.7.0                      py_0    conda-forge
cryptography              2.4.2            py37ha12b0ac_0
cycler                    0.10.0                     py_1    conda-forge
cytoolz                   0.9.0.1          py37h470a237_1    conda-forge
dask                      1.1.1                      py_0    conda-forge
dask-core                 1.1.1                      py_0    conda-forge
decorator                 4.3.2                      py_0    conda-forge
dill                      0.2.9                    py37_0    conda-forge
distributed               1.25.3                   py37_0    conda-forge
entrypoints               0.3                   py37_1000    conda-forge
fastcache                 1.0.2           py37h1de35cc_1001    conda-forge
freetype                  2.9.1             h597ad8a_1005    conda-forge
gmp                       6.1.2             h0a44026_1000    conda-forge
gmpy2                     2.0.8           py37h3aece09_1002    conda-forge
h5py                      2.9.0           nompi_py37h35dc34b_1102    conda-forge
hdf5                      1.10.4          nompi_h646315f_1105    conda-forge
heapdict                  1.0.0                 py37_1000    conda-forge
hyperspy                  1.4.1                 py37_1002    conda-forge
hyperspy-base             1.4.1           py37h1de35cc_1002    conda-forge
hyperspy-gui-ipywidgets   1.1.1                      py_1    conda-forge
hyperspy-gui-traitsui     1.1.1                 py37_1000    conda-forge
icu                       58.2              h0a44026_1000    conda-forge
idna                      2.8                   py37_1000    conda-forge
imageio                   2.4.1                 py37_1000    conda-forge
intel-openmp              2019.1                      144
ipykernel                 5.1.0           py37h24bf2e0_1002    conda-forge
ipyparallel               6.2.3                 py37_1000    conda-forge
ipython                   7.2.0           py37h24bf2e0_1000    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
ipywidgets                7.4.2                      py_0    conda-forge
jedi                      0.13.2                py37_1000    conda-forge
jinja2                    2.10                       py_1    conda-forge
jpeg                      9c                h1de35cc_1001    conda-forge
jsonschema                3.0.0a3               py37_1000    conda-forge
jupyter                   1.0.0                      py_1    conda-forge
jupyter_client            5.2.4                      py_1    conda-forge
jupyter_console           6.0.0                      py_0    conda-forge
jupyter_core              4.4.0                      py_0    conda-forge
jupyterlab                0.35.4                   py37_0    conda-forge
jupyterlab_server         0.2.0                      py_0    conda-forge
kiwisolver                1.0.1           py37h04f5b5a_1002    conda-forge
libcxx                    4.0.1                hcfea43d_1
libcxxabi                 4.0.1                hcfea43d_1
libedit                   3.1.20181209         hb402a30_0
libffi                    3.2.1                h475c297_4
libgfortran               3.0.1                h93005f0_2
libpng                    1.6.36            ha441bb4_1000    conda-forge
libsodium                 1.0.16            h1de35cc_1001    conda-forge
libtiff                   4.0.10            h79f4b77_1001    conda-forge
link-traits               1.0.2                 py37_1002    conda-forge
llvmlite                  0.25.0          py37h3fea490_1001    conda-forge
locket                    0.2.0                      py_2    conda-forge
markupsafe                1.1.0           py37h1de35cc_1000    conda-forge
matplotlib                3.0.2                 py37_1002    conda-forge
matplotlib-base           3.0.2           py37hf043ca5_1002    conda-forge
mistune                   0.8.4           py37h1de35cc_1000    conda-forge
mkl                       2019.1                      144
mkl_fft                   1.0.10           py37h1de35cc_1    conda-forge
mkl_random                1.0.2            py37h1702cab_2    conda-forge
mpc                       1.1.0             h3aece09_1006    conda-forge
mpfr                      4.0.1             h0a95893_1000    conda-forge
mpmath                    1.1.0                      py_0    conda-forge
msgpack-python            0.6.1            py37h04f5b5a_0    conda-forge
natsort                   6.0.0                      py_0    conda-forge
nbconvert                 5.3.1                      py_1    conda-forge
nbformat                  4.4.0                      py_1    conda-forge
ncurses                   6.1                  h0a44026_1
networkx                  2.2                        py_1    conda-forge
nodejs                    11.9.0               h0a44026_0    conda-forge
notebook                  5.7.4                 py37_1000    conda-forge
numba                     0.40.1          py37h1702cab_1000    conda-forge
numexpr                   2.6.9           py37h1702cab_1000    conda-forge
numpy                     1.15.4           py37hacdab7b_0
numpy-base                1.15.4           py37h6575580_0
olefile                   0.46                       py_0    conda-forge
openssl                   1.1.1a            h1de35cc_1000    conda-forge
packaging                 19.0                       py_0    conda-forge
pandas                    0.24.1           py37h0a44026_0    conda-forge
pandoc                    2.6                           0    conda-forge
pandocfilters             1.4.2                      py_1    conda-forge
parso                     0.3.2                      py_0    conda-forge
partd                     0.3.9                      py_0    conda-forge
patsy                     0.5.1                      py_0    conda-forge
pexpect                   4.6.0                 py37_1000    conda-forge
pickleshare               0.7.5                 py37_1000    conda-forge
pillow                    5.3.0           py37hbddbef0_1000    conda-forge
pint                      0.9                      py37_2    conda-forge
pip                       18.1                     py37_0
prometheus_client         0.5.0                      py_0    conda-forge
prompt_toolkit            2.0.8                      py_0    conda-forge
psutil                    5.5.0            py37h1de35cc_0    conda-forge
ptyprocess                0.6.0                 py37_1000    conda-forge
pycparser                 2.19                       py_0    conda-forge
pyface                    6.0.0                      py_1    conda-forge
pygments                  2.3.1                      py_0    conda-forge
pyopenssl                 19.0.0                   py37_0    conda-forge
pyparsing                 2.3.1                      py_0    conda-forge
pyqt                      5.6.0           py37hc26a216_1008    conda-forge
pyrsistent                0.14.9          py37h1de35cc_1000    conda-forge
pysocks                   1.6.8                 py37_1002    conda-forge
python                    3.7.2                haf84260_0
python-dateutil           2.7.5                      py_0    conda-forge
pytz                      2018.9                     py_0    conda-forge
pywavelets                1.0.1           py37h917ab60_1000    conda-forge
pyyaml                    3.13            py37h1de35cc_1001    conda-forge
pyzmq                     17.1.2          py37h111632d_1001    conda-forge
qt                        5.6.2             h822fa55_1013    conda-forge
qtconsole                 4.4.3                      py_0    conda-forge
readline                  7.0                  h1de35cc_5
requests                  2.21.0                py37_1000    conda-forge
scikit-image              0.14.2           py37h0a44026_1    conda-forge
scikit-learn              0.20.2           py37h27c97d8_0
scipy                     1.2.0            py37h1410ff5_0
send2trash                1.5.0                      py_0    conda-forge
setuptools                40.6.3                   py37_0
sip                       4.18.1          py37h0a44026_1000    conda-forge
six                       1.12.0                py37_1000    conda-forge
sortedcontainers          2.1.0                      py_0    conda-forge
sparse                    0.6.0                      py_0    conda-forge
sqlite                    3.26.0               ha441bb4_0
statsmodels               0.9.0           py37h917ab60_1000    conda-forge
sympy                     1.3                   py37_1000    conda-forge
tblib                     1.3.2                      py_1    conda-forge
terminado                 0.8.1                 py37_1001    conda-forge
testpath                  0.4.2                 py37_1000    conda-forge
tk                        8.6.8                ha441bb4_0
toolz                     0.9.0                      py_1    conda-forge
tornado                   5.1.1           py37h1de35cc_1000    conda-forge
tqdm                      4.30.0                     py_0    conda-forge
traitlets                 4.3.2                 py37_1000    conda-forge
traits                    5.0.0            py37h1de35cc_0    conda-forge
traitsui                  6.0.0                      py_1    conda-forge
urllib3                   1.24.1                py37_1000    conda-forge
wcwidth                   0.1.7                      py_1    conda-forge
webencodings              0.5.1                      py_1    conda-forge
wheel                     0.32.3                   py37_0
widgetsnbextension        3.4.2                 py37_1000    conda-forge
xz                        5.2.4                h1de35cc_4
yaml                      0.1.7             h1de35cc_1001    conda-forge
zeromq                    4.2.5             h0a44026_1006    conda-forge
zict                      0.1.3                      py_0    conda-forge
zlib                      1.2.11               h1de35cc_3
```

**Comment by ZGainsforth** [comment]
It's worth mentioning that I had to do something similar with fourier_ratio_deconvolution().  The code as follows: 

```
import numpy as np
import matplotlib.pyplot as plt
import hyperspy.api as hs
import os
import matplotlib
print(matplotlib.rcParams['backend'])

ZLP = hs.load('007-EELS-ZLP-45pA.emi')
ZLP.set_signal_type('EELS')
ZLP.set_microscope_parameters(beam_energy=300, convergence_angle=2, collection_angle=30)
FeL = hs.load('008-EELS_Map_core-360pA.emi')
FeL.set_signal_type('EELS')
FeL.set_microscope_parameters(beam_energy=300, convergence_angle=2, collection_angle=30)
print(f'ZLP shape is: {ZLP.data.shape}')
print(f'FeL shape is: {FeL.data.shape}')

ZLP.align_zero_loss_peak()

FeSub = FeL.remove_background(signal_range=(670., 695.), background_type='PowerLaw', fast=False)

# -------------NOTE HERE!!!!!!! ---------------
# THIS LINE FAILS
FeDeconv = FeSub.fourier_ratio_deconvolution(ll=ZLP) 
# THIS LINE WORKS (if uncommented, obviously)
# FeDeconv = FeSub.fourier_ratio_deconvolution(ll=ZLP, extrapolate_lowloss=False, extrapolate_coreloss=False) 
# -------------NOTE HERE!!!!!!! ---------------

ZLP.plot()
FeL.plot()
FeSub.plot()
plt.show()
```

fails with:

```
/Users/Zack/Git/hyperspy/hyperspy/_components/power_law.py:135: RuntimeWarning: invalid value encountered in log
  r = 2 * log(I1 / I2) / log(x2 / x1)
/Users/Zack/Git/hyperspy/hyperspy/_components/power_law.py:137: RuntimeWarning: overflow encountered in power
  A = k * I2 / (x2 ** k - x3 ** k)
/Users/Zack/Git/hyperspy/hyperspy/_components/power_law.py:137: RuntimeWarning: invalid value encountered in subtract
  A = k * I2 / (x2 ** k - x3 ** k)
WARNING:hyperspy._components.power_law:Power law paramaters estimation failed because of a "divide by zero" error.
Traceback (most recent call last):
  File "ReproduceBug.py", line 20, in <module>
    FeDeconv = FeSub.fourier_ratio_deconvolution(ll=ZLP)
  File "/Users/Zack/Git/hyperspy/hyperspy/_signals/eels.py", line 725, in fourier_ratio_deconvolution
    extrapolation_size=100)
  File "/Users/Zack/Git/hyperspy/hyperspy/_signals/eels.py", line 975, in power_law_extrapolation
    out=True)
TypeError: cannot unpack non-iterable bool object
```

It is entirely possible there is a bad pixel in the map or something which is exposing the bug, but it isn't handling it gracefully.  As long as I tell it not to extrapolate the power law on the spectra it works fine, however.

Hopefully this is useful info too.

**Comment by ericpre** [comment]
> When I take out the signal_range in an attempt to activate interactive mode it seems to need 'Power Law' instead of 'PowerLaw' and even then errors out. So perhaps I'm not doing that part right:

Yes, indeed you need to use `"Power Law"` and not `"PowerLaw"` as described in the docstring.

> `FeSub = FeL.remove_background(background_type='Power Law', fast=False)`
> 
> results in:
> 
>```
>VBox(children=(FloatText(value=nan, description='Left', disabled=True), FloatText(value=nan, description='Right', disabled=True), Dropdown(description='Background type', options=('Power Law', 'Gaussian', 'Offset', 'Polynomial'), value='Power Law'), IntText(value=2, description='Polynomial order', layout=Layout(display='none')), Checkbox(value=False, description='Fast'), Checkbox(value=False, description='Zero Fill'), Accordion(children=(HTML(value='Click on the signal figure and drag to the right to select a range. Press `Apply` to remove the background in the whole dataset. If "Fast" is checked, the background parameters are estimated using a fast (analytical) method that can compromise accuracy. When unchecked, non-linear least squares is employed instead. If "Zero Fill" is checked, all the channels prior to the fitting region will be set to zero. Otherwise the background subtraction will be performed in the pre-fitting region as well.'),), _titles={'0': 'Help'}), HBox(children=(Button(description='Apply', style=ButtonStyle(), tooltip='Remove the background in the whole dataset.'), Button(description='Close', style=ButtonStyle(), tooltip='Close widget and remove span selector from the signal figure.')))))
> Traceback (most recent call last):
>   File "ReproduceBug.py", line 25, in <module>
>     FeSub.plot()
> AttributeError: 'NoneType' object has no attribute 'plot'
> ```

This is normal (I was a bit confuse between the different issue here...), if you use `signal_range=interactive"`, it removes the background in place and returns `None`, therefore `FeSub` is `Nonetype`.

**Comment by ericpre** [comment]
The polynomial background issue is fixed in #1989.

**Comment by ericpre** [comment]
Fixed in #1989.
